### PR TITLE
Replication initial checkin

### PR DESCRIPTION
**Pull request by bjeevan-ib**
pgctl implements the replication logic.
now the controller supports newDB, existingDB and migration from existingDB to newDB.

### CODE REVIEW

File: controllers/databaseclaim_controller.go [hunk 1]
```diff
@@ -465,132 +480,59 @@ func (r *DatabaseClaimReconciler) isResourceReady(resourceStatus xpv1.ResourceSt
...
-	// Since update does not work I am not going to make this change as it will impact
-	// the performance of the controller.
-	//
-	err := r.Client.Get(ctx, client.ObjectKey{
-		Name: dbHostName,
-	}, dbInstance)
-	if err != nil {
-		err = r.createCloudDatabase(dbHostName, ctx, fragmentKey, dbClaim)
+	if dbClaim.Spec.Type == defaultPostgresStr {
+		return r.createPostgresDBInstanceCR(ctx, dbHostIdentifier, dbClaim)
```
Comment (abalaven) [inline controllers/databaseclaim_controller.go:525]: Rename to managePostgresDBInstance

File: controllers/databaseclaim_controller.go [hunk 2]
```diff
@@ -465,132 +480,59 @@ func (r *DatabaseClaimReconciler) isResourceReady(resourceStatus xpv1.ResourceSt
...
+	} else if dbClaim.Spec.Type == defaultAuroraPostgresStr {
+		ready, err := r.createDBClusterCR(ctx, dbHostIdentifier, dbClaim)
```
Comment (abalaven) [inline controllers/databaseclaim_controller.go:527]: Rename to manageDBCluster

File: controllers/databaseclaim_controller.go [hunk 3]
```diff
@@ -465,132 +480,59 @@ func (r *DatabaseClaimReconciler) isResourceReady(resourceStatus xpv1.ResourceSt
...
-
-	// Check if resource is ready
-	if !r.isResourceReady(dbInstance.Status.ResourceStatus) {
-		return connInfo, nil
-	}
-	if dbClusterRequired {
-		if !r.isResourceReady(dbCluster.Status.ResourceStatus) {
-			return connInfo, nil
+		if ready {
+			return r.createAuroraDBInstanceCR(ctx, dbHostIdentifier, dbClaim)
```
Comment (abalaven) [inline controllers/databaseclaim_controller.go:532]: Rename to manageAuroraDBInstance

File: controllers/databaseclaim_controller.go [hunk 4]
```diff
@@ -659,95 +595,38 @@ func (r *DatabaseClaimReconciler) getDynamicHostParams(ctx context.Context, frag
...
-		Key: "secret",
-	}
-
 	dbMasterSecretCluster := xpv1.SecretKeySelector{
 		SecretReference: xpv1.SecretReference{
-			Name:      dbHostName + "-master-cluster",
+			Name:      dbHostName + "-master",
 			Namespace: serviceNS,
 		},
 		Key: "secret",
```
Comment (abalaven) [inline controllers/databaseclaim_controller.go:616]:
```suggestion
		Key: "password",
```

File: controllers/databaseclaim_controller.go [hunk 5]
```diff
@@ -785,98 +664,245 @@ func (r *DatabaseClaimReconciler) createCloudDatabase(dbHostName string, ctx con
...
-		r.Log.Info("unable delete crossplane DBInstance resource", "DBInstance", dbHostName)
-		return err
-	} else {
-		r.Log.Info("deleted crossplane DBInstance resource", "DBInstance", dbHostName)
+	dbMasterSecretInstance := xpv1.SecretKeySelector{
+		SecretReference: xpv1.SecretReference{
+			Name:      dbHostName + "-master",
+			Namespace: serviceNS,
+		},
+		Key: "secret",
```
Comment (abalaven) [inline controllers/databaseclaim_controller.go:696]:
```suggestion
		Key: "password",
```

File: controllers/databaseclaim_controller.go [hunk 6]
```diff
@@ -785,98 +664,245 @@ func (r *DatabaseClaimReconciler) createCloudDatabase(dbHostName string, ctx con
...
-				ResourceSpec: xpv1.ResourceSpec{
-					WriteConnectionSecretToReference: &dbSecretInstance,
-					ProviderConfigReference:          &providerConfigReference,
-					DeletionPolicy:                   params.DeletionPolicy,
-				},
-			},
+		} else {
+			return false, err
 		}
 	}
```
Comment (abalaven) [inline controllers/databaseclaim_controller.go:670]: Add check to make sure deletionTimestamp is not 0

File: controllers/databaseclaim_controller.go [hunk 7]
```diff
@@ -1109,19 +1132,132 @@ func (r *DatabaseClaimReconciler) manageSuccess(ctx context.Context, dbClaim *pe
...
+
+func getServiceNamespace() (string, error) {
+	ns, found := os.LookupEnv(serviceNamespaceEnvVar)
+	if !found {
+		return "", fmt.Errorf("service namespace env %s must be set", serviceNamespaceEnvVar)
+	}
+	return ns, nil
+}
+
+func (r *DatabaseClaimReconciler) reconcileUseExisitngDB(ctx context.Context, dbClaim *persistancev1.DatabaseClaim) error {
```
Comment (abalaven) [inline controllers/databaseclaim_controller.go:1217]: nit: reconcileUseExistingDB

File: controllers/databaseclaim_controller.go (hunk 1 above)
Comment (bjeevan-ib) [reply controllers/databaseclaim_controller.go:525]: done

File: controllers/databaseclaim_controller.go (hunk 6 above)
Comment (bjeevan-ib) [reply controllers/databaseclaim_controller.go:670]: done

**Review by abalaven — APPROVED** [review]
LGTM, thanks for addressing the comments!

**Pull request merged**